bf: S3C-3061 allow bucket acl via predefined groups

naren-rajendran opened the pull request:
Corrections to bucket ACL using s3 predefined groups. 

Ref: https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html

Commit:
bf: S3C-3061 allow bucket acl via predefined groups

bert-e commented:
# Hello naren-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/cloudserver/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups` with contents from `bugfix/S3C-3061-bucket-acl-using-predefined-groups`
and `development/7.10`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups origin/development/7.10
 $ git merge origin/bugfix/S3C-3061-bucket-acl-using-predefined-groups
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups
```

naren-rajendran commented:
Note to reviewers, the tests have to be refactored.

miniscruff commented on tests/functional/aws-node-sdk/test/bucket/aclUsingPredefinedGroups.js:
```diff
@@ -0,0 +1,634 @@
+    }
+
+    // tests for authenticated user(signed) and anonymous user(unsigned)
+    [true, false].forEach(auth => {
```
splitting all these tests with a bool here is making it quite hard to read and adds at least one if check to every test, I think it would be better to create two describes instead.

naren-rajendran replied:
The tests are the same for both the cases and that is why I had to do this. I originally had separate tests.

miniscruff replied:
I would not consider the tests the same if you need an if/else in the test based on the condition. It is not like a table test where you are testing many inputs against the same method and output. `auth` is used as a conditional check a lot.

naren-rajendran replied:
Refactored a bit.

miniscruff replied:
I will accept it, but I tend to dislike conditionals in tests as I think a straight top-down given-when-then flow works more consistently.

miniscruff commented on tests/functional/aws-node-sdk/test/bucket/aclUsingPredefinedGroups.js:
```diff
@@ -0,0 +1,634 @@
+                }, err => {
+                    assert.ifError(err);
+                    const param = { Bucket: testBucket };
+                    const cb = er => {
+                        assert.ifError(er);
+                        done();
+                    };
```
Instead of including this each time you could create a helper method

```js
function wrapCallback(done) {
  return err => {
    assert.ifError(err);
    done();
  };
}
```
Then use `wrapCallback(done)` in place of `cb` in all successful future cases.

miniscruff commented on tests/functional/aws-node-sdk/test/bucket/aclUsingPredefinedGroups.js:
```diff
@@ -0,0 +1,634 @@
+                        Key: testKey,
+                    };
+                    const cb = er => {
+                        assert(er.code, errors.AccessDenied);
```
Did you mean to use `assert.strictEqual`? This will return true if er.code is not 0, then use access denied as the message.

naren-rajendran replied:
Yep, it was a copy-paste.

jonathan-gramain commented on lib/api/objectDelete.js:
```diff
@@ -73,6 +69,12 @@ function objectDelete(authInfo, request, log, cb) {
                 }
+                // based on AWS's behavior, object delete can be performed by
+                // either bucket owner or the object owner
+                if (objMD['owner-id'] !== canonicalID
```
The difference is subtle, but if you move this check above the `if (!objMD)` and add the `objMD && ` condition at the beginning (similar to what is done for `objectPut`), it allows to return `AccessDenied` whether the object exists or not, without leaking info about the object existence for unauthorized users. I don't know what AWS does in this respect but it could be good to check.

naren-rajendran replied:
Yes, corrected. Checked AWS behavior, you get access denied on non existing object.

naren-rajendran commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups` with contents from `bugfix/S3C-3061-bucket-acl-using-predefined-groups`
and `development/7.10`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups origin/development/7.10
 $ git merge origin/bugfix/S3C-3061-bucket-acl-using-predefined-groups
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups
```

naren-rajendran commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups` with contents from `bugfix/S3C-3061-bucket-acl-using-predefined-groups`
and `development/7.10`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups origin/development/7.10
 $ git merge origin/bugfix/S3C-3061-bucket-acl-using-predefined-groups
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups
```

bert-e commented:
# History mismatch



Merge commit #9b36b19a5a21feb0069ae52757db918b1aa1ede9 on the integration branch
`w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups` is merging a branch which is neither the current
branch `bugfix/S3C-3061-bucket-acl-using-predefined-groups` nor the development branch
`development/7.10`.

It is likely due to a rebase of the branch `bugfix/S3C-3061-bucket-acl-using-predefined-groups` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

naren-rajendran commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups` with contents from `bugfix/S3C-3061-bucket-acl-using-predefined-groups`
and `development/7.10`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups origin/development/7.10
 $ git merge origin/bugfix/S3C-3061-bucket-acl-using-predefined-groups
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups
```

bert-e commented:
# History mismatch



Merge commit #ec35a60ee10cec180511ce47a0fb369ca2c2abeb on the integration branch
`w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups` is merging a branch which is neither the current
branch `bugfix/S3C-3061-bucket-acl-using-predefined-groups` nor the development branch
`development/7.10`.

It is likely due to a rebase of the branch `bugfix/S3C-3061-bucket-acl-using-predefined-groups` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

naren-rajendran commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups` with contents from `bugfix/S3C-3061-bucket-acl-using-predefined-groups`
and `development/7.10`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups origin/development/7.10
 $ git merge origin/bugfix/S3C-3061-bucket-acl-using-predefined-groups
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups` with contents from `w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups`
and `development/8.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups origin/development/8.2
 $ git merge origin/w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/S3C-3061-bucket-acl-using-predefined-groups` into
`development/7.4`
* [w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups](https://github.com/scality/cloudserver/compare/development/7.10...w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups) will be merged into `development/7.10` (pull request #3615)
* [w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups](https://github.com/scality/cloudserver/compare/development/8.2...w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups) will be merged into `development/8.2` (pull request #3616)



The following branches will **NOT** be impacted:

* `development/6.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
bf: S3C-3061 revert writes acl fix

reverting writes acl fix as this conflicts with service accounts
behavior, which will be handled separately.

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups` with contents from `w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups`
and `development/8.2`.

Please resolve the conflict on **the integration branch** (`w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups
 $ git pull  # or "git reset --hard origin/w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups"
 $ git merge origin/development/8.2
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups
```

naren-rajendran commented:
@bert-e reset

bert-e commented:
# Lossy reset warning



There seems to be *manual* commits on integration branches (e.g. conflict
resolutions) that will be lost if you chose to *reset*.

**You can use the `force_reset` command if you still want me
to delete those branches.**

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups` with contents from `w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups`
and `development/8.2`.

Please resolve the conflict on **the integration branch** (`w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups
 $ git pull  # or "git reset --hard origin/w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups"
 $ git merge origin/development/8.2
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups
```

naren-rajendran commented:
@bert-e force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups` with contents from `bugfix/S3C-3061-bucket-acl-using-predefined-groups`
and `development/7.10`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups origin/development/7.10
 $ git merge origin/bugfix/S3C-3061-bucket-acl-using-predefined-groups
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/S3C-3061-bucket-acl-using-predefined-groups` into
`development/7.4`
* [w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups](https://github.com/scality/cloudserver/compare/development/7.10...w/7.10/bugfix/S3C-3061-bucket-acl-using-predefined-groups) will be merged into `development/7.10` (pull request #3617)
* [w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups](https://github.com/scality/cloudserver/compare/development/8.2...w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups) will be merged into `development/8.2` (pull request #3618)



The following branches will **NOT** be impacted:

* `development/6.4`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

naren-rajendran commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/cloudserver/commit/e14a6e74fb5ab8e1c4bcfdca061320feb8196b34) did not succeed in branch w/8.2/bugfix/S3C-3061-bucket-acl-using-predefined-groups.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.4`

* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/8.2`



The following branches will **NOT** be impacted:

* `development/6.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.4`

* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/6.4`



Please check the status of the associated issue S3C-3061.

Goodbye naren-scality.
